src/xrpld/core/detail/DatabaseCon.cpp

```diff
@@ -237,6 +237,55 @@ setup_DatabaseCon(Config const& c, std::optional<beast::Journal> j)
     }
     setup.useGlobalPragma = true;
 
+    auto setPragma =
+        [](std::string& pragma, std::string const& key, int64_t value) {
+            pragma = "PRAGMA " + key + "=" + std::to_string(value) + ";";
+        };
+
+    auto isPageSizeValid = [](int64_t page_size) {
+        return (page_size >= 512 && page_size <= 65536);
+    };
+
+    auto isPageSizePowerOfTwo = [](int64_t page_size) {
+        return (page_size > 0) && ((page_size & (page_size - 1)) == 0);
+    };
+
+    // Lgr Pragma
+    setPragma(setup.lgrPragma[0], "journal_size_limit", 1582080);
+
+    // TX Pragma
+    int64_t page_size = 4096;
+    int64_t journal_size_limit = 1582080;
+    int64_t max_page_count = 4294967294;
+
+    if (c.exists("sqlite"))
+    {
+        auto& s = c.section("sqlite");
+        set(journal_size_limit, "journal_size_limit", s);
+        set(page_size, "page_size", s);
+        if (!isPageSizeValid(page_size))
+            Throw<std::runtime_error>(
+                "Invalid page_size. Must be between 512 and 65536.");
+
+        if (!isPageSizePowerOfTwo(page_size))
+            Throw<std::runtime_error>(
+                "Invalid page_size. Must be a power of 2.");
+    }
+
+    setPragma(setup.txPragma[0], "page_size", page_size);
+    setPragma(setup.txPragma[1], "journal_size_limit", journal_size_limit);
+    setPragma(setup.txPragma[2], "max_page_count", max_page_count);
+
+#if (ULONG_MAX > UINT_MAX) && !defined(NO_SQLITE_MMAP)
+    setup.txPragma[3] = "PRAGMA mmap_size=17179869184;";
+#else
+    // Provide an explicit `no-op` SQL statement
```

Comment:
I'd prefer is we switched the type of `setup.txPragmas` to `std::vector<std::string>` rather than use a dummy statement.

Reply:
I think the reason why we used `std::array` was for `constexpr` initialisation inside `DBInit.h`; this is out of the window, now that we are building the set of init pragmas in runtime, so there is no point sticking to the limitations of `std::array` either.

Reply:
Actually, we can keep `std::array<std::string>` ; the whole conditional is pointless and we can remove the `#ifdef #else #endif`.

We can safely assume that the `#if` condition is true because we do not support 32bit builds; the `NO_SQLITE_MMAP` is never passed by the cmake files and this would only matter if someone was trying to use version of sqlite3 older than [version 3.7.17, released in 2013](https://www.sqlite.org/releaselog/3_7_17.html).

Reply:
Thank you for taking the time to look that up and review.